tmp.ts

```diff
@@ -0,0 +1,157 @@
+   */
+  async linkAddress(address: string, provider: any): Promise<LinkProof> {
+    if (!this._keyring)
+      throw new Error('This method can only be called after authenticate has been called')
+    const proof = await createLink(this.DID!, address, provider)
+    this.events.emit('new-link-proof', proof)
+    return proof
+  }
+
+  async linkManagementKey() {
+    if (this._externalAuth) return null
+    const timestamp = Math.floor(new Date().getTime() / 1000)
+    const msg = `Create a new 3Box profile\n\n- \nYour unique profile ID is ${this
+      .DID!} \nTimestamp: ${timestamp}`
+    return {
+      msg,
+      timestamp,
+      sig: await this._keyring!.managementPersonalSign(msg),
+    }
+  }
+
+  async _initKeyring(authData?: Array<EncryptedMessage>, address?: string, spaces?: Array<string>) {
+    if (this._seed) {
+      await this.getLink()
+    } else if (authData && authData.length > 0) {
+      let seed
+      authData.find(({ ciphertext, nonce }) => {
+        seed = Keyring.decryptWithAuthSecret(ciphertext, nonce, this._authSecret as string)
+        return Boolean(seed)
+      })
+      if (!seed) throw new Error('No valid auth-secret for this identity')
+      this._keyring = new Keyring(seed)
+      this.DID = await this._get3id()
+    } else if (this._externalAuth) {
+      if (!address) throw new Error('External authentication requires an address')
+      const migratedKeys = await this._externalAuth({
+        address,
+        spaces,
+        type: '3id_migration',
+      })
+      this._keyring = new Keyring(null, migratedKeys)
+      this.DID = await this._get3id()
+      const proof = await this._externalAuth({
+        address,
+        type: '3id_createLink',
+        did: this.DID,
+      })
+      if (proof) this.events.emit('new-link-proof', proof)
+    } else {
+      // no authData available so we create a new identity
+      const seed = '0x' + Buffer.from(naclRandom(32)).toString('hex')
+      this._keyring = new Keyring(seed)
+      this.DID = await this._get3id()
+      await this.addAuthMethod(this._authSecret!)
+    }
+  }
+
+  /**
+   * Authenticate to given spaces
+   *
+   * @param     {Array<String>}     spaces          The desired spaces
+   * @param     {Object}    opts                    Optional parameters
+   * @param     {Array<Object>}     opts.authData   The authData for this identity
+   * @return    {Object}                            The public keys for the requested spaces of this identity
+   */
+  async authenticate(
+    spaces: Array<string> = [],
+    {
+      authData,
+      address,
+      mgmtPub,
+    }: {
+      authData?: Array<EncryptedMessage>
+      address?: string
+      mgmtPub?: string
+    } = {},
+    origin?: string | null
+  ): Promise<{ main: PublicKeys; spaces: Record<string, PublicKeys> }> {
+    if (!this._keyring) throw new Error('Can not authenticate yet')
+    let permission = await this.getPermission(spaces, origin, { address })
```

Comment:
Not an issue but the `permission` value doesn't seem to get changed so it could be assigned using `const`

Reply:
Err, ty. This file should not be here.